### ISSUE TITLE
Parameters sorted incorrectly inside nginx.conf

### ISSUE THREAD

**Issue by TaiSHiNet** [saltstack-formulas/nginx-formula#40]
Currently I'm suffering an issue due to contents being sorted with includes first:

```
http {
    access_log /var/log/nginx/access.log;
    client_max_body_size 10m;
    default_type application/octet-stream;
    error_log /var/log/nginx/error.log;
    gzip off;
    gzip_disable "msie6";

    include /etc/nginx/mime.types;
    include /etc/nginx/conf.d/*.conf;
    include /etc/nginx/sites-enabled/*;
    keepalive_timeout 65;
    proxy_cache_path /var/run/nginx-cache keys_zone=temp:100m;
    proxy_temp_path /var/run/nginx-tmp;
    sendfile on;
    server_names_hash_bucket_size 128;
    tcp_nodelay on;
    tcp_nopush on;
    types_hash_max_size 2048;
    underscores_in_headers on;
}
```

Given that I'm trying to use cache, this makes my configuration fail (vhost is loaded before cache is defined)

**Comment by Seldaek** [comment]
I just hit the same issue with `fastcgi_cache_path` and `fastcgi_cache` both defined in http{} but when sorted the cache comes first and creates a conflict when cache_path tries to define the cache name.

I could solve it by moving `fastcgi_cache` into the server{} block in my vhosts, but that's just lucky because fastcgi sorts before include so definitely would be happy to see this fixed as well.

**Comment by kurt---** [comment]
+1 for this issue. From my perspective it is impossible to configure an upstream proxy because the vhost is always included before the cache_path or the cache directive is printed before the cache_path.

**Comment by TaiSHiNet** [comment]
My current workaround is creating a vhost called 000-whatever, but I'd love to have this fixed

**Comment by kurt---** [comment]
I also made a workaround which works for me:

https://github.com/kurt---/nginx-formula/tree/enable-proxy

it works by adding a proxy.conf-configuration-file which is inserted before the proxy_cache-directive. The config-file enables the cache (and logging-option) and one can enable the cache if desired. to enable the additional access-log one need to put these into the vhost-config else we ran into the same problem with options in the wrong order. 

So this is not a fix but it works for me to enable a proxy in nginx.

**Comment by outime** [comment]
Is this still broken?

**Comment by gravyboat** [comment]
@outime This is specifically for the ng formula as far as I'm aware, are you using that portion?

**Comment by aboe76** [comment]
@gravyboat it still broken only @EvaSDK workaround works.

**Comment by puneetk** [comment]
@EvaSDK or @aboe76 can you send a pull request with the work around please, we ll get it into mainline for the users of the formula.

**Comment by aboe76** [comment]
@EvaSDK thanks, for this, I now it doesn't look nice but it works.

**Comment by EvaSDK** [comment]
yeah, I hope someone finds a better workaround but at least it gets it working for some use cases.

**Comment by bentwire** [comment]
This also affects setting the logging format as well.  Due to log_format being defined in the config after access_log.

**Comment by ernstae** [comment]
This is definitely causing a problem in Ubuntu for me.  We are using the streaming module for TCP load balancing in a few of our environments.
Currently, the workaround is to hackily do this:

```
+# this is a hack because it won't add correctly in the pillar (gets re-ordereed)
+include /etc/nginx/modules-enabled/*.conf;
```
making sure this is on the first (or nearly first) line of the nginx configuration, because without it the `stream{}` directive is unable to load.

This reared its ugly head tonight, because we rebased with your repo, and the hacky customization got missed in the internal pull request on our saltstack repo.

Any chance this use case could be handled?

**Comment by ernstae** [comment]
I may have done a less hacky way to workaround my above comment, by adding another include to the pillar, which is at least a sustainable way around:

```nginx:
    ng:
      server:
        config:
          include: '/etc/nginx/modules-enabled/*.conf'
          stream:
            server:
              listen: 7999```
